src/emulator/functionsEmulator.ts

```diff
@@ -432,23 +448,28 @@ export class FunctionsEmulator implements EmulatorInstance {
    *
    * TODO(abehaskins): Gracefully handle removal of deleted function definitions
    */
-  async loadTriggers(force = false): Promise<void> {
+  async loadTriggers(emulatableBackend: EmulatableBackend, force = false): Promise<void> {
     // Before loading any triggers we need to make sure there are no 'stale' workers
     // in the pool that would cause us to run old code.
     this.workerPool.refresh();
```

Comment:
I'm worried about this as well, and its very possible that I'll run into some subtle bugs with this once we start loading code from multiple places. However, i did some hacking around locally + code sleuthing to try to validate my concerns, and I _think_ this is safe to do, if probably inefficient.

 workerPool.refresh() kills all the child processes that we already created that are not currently doing work, and marks the ones that are doing work for later cleanup. Therefore, subsequent calls to loadTriggers won't interrupt others that are running. They also each listen to just one process for parsedTriggers, so I don't believe they'll pick up triggerParseEvents from other invocations. The biggest problem i can see now is that we clean up _all_ old workers, even ones that are running code for different functions that aren't affected by this loadTriggers. This means more process churning and worse performance.

If I have time in the future, I'd consider trying to make workerPool.refresh only clear out workers running code from the same functionsDir - however, for now, I'm ok with sacrificing some performance for simplicity.

Reply:
1 - We may want to turn it into a map (or rethink this pattern entirely), but I don't believe we _need_ to in order to support multiple runtimes. The worker pool already has logic for matching a triggerId to a worker for that trigger (https://github.com/firebase/firebase-tools/blob/master/src/emulator/functionsRuntimeWorker.ts#L242). I don't see any reason that inherently can't work with workers that handle different runtimes

2- Yep, I'm sure it only soft shutdowns active workers. Here is where we mark them as FINISHING instead of FINISHED (https://github.com/firebase/firebase-tools/blob/a90f28aacb23618fd7101960b11ab0e421c25072/src/emulator/functionsRuntimeWorker.ts#L185), and then the next time the worker emits a runtime-status log, it kills the process (https://github.com/firebase/firebase-tools/blob/a90f28aacb23618fd7101960b11ab0e421c25072/src/emulator/functionsRuntimeWorker.ts#L48).